#659: Add support for command `LINSERT` and `LRANGE`

ParvBudh28 opened the pull request:
* Adds support for command `LINSERT` and `LRANGE`.
* Adds `DequeIterator` for traversing over the `deque`. `LRANGE` command should now be straight forward to implement using the iterator.

Closes #659

ParvBudh28 commented:
@JyotinderSingh Can you please review this or assign an appropriate reviewer for this.

TIA

JyotinderSingh commented:
There seem to be some conflicts, could you please rebase on the latest master?

ParvBudh28 commented:
> There seem to be some conflicts, could you please rebase on the latest master?

Done.

ParvBudh28 commented:
@JyotinderSingh @apoorvyadav1111 Can you please review this PR.

TIA

JyotinderSingh commented on integration_tests/commands/async/deque_test.go:
```diff
@@ -442,6 +442,40 @@ func TestLLEN(t *testing.T) {
+	}{
+		{
+			name:   "LINSERT before",
+			cmds:   []string{"LPUSH k v1 v2 v3 v4", "LINSERT k before v2 e1", "LINSERT k before v1 e2", "LINSERT k before v4 e3"},
```
The test case should also verify the final state of the list by running `LRANGE` at the end.

ParvBudh28 replied:
We don't have LRANGE command implemented yet. I didn't want to add it as a part of this PR since it will get too long. It's pretty straight forward to implement LRANGE after the introduction of the iterator in this PR. Let's create a new issue request for this and update these tests then. For now I can add a TODO item here. 

Let me know your thoughts here.

JyotinderSingh replied:
I see, that sounds good. Please add a todo, additionally, it would be good to create an issue as well for the same to allow for better tracking of work. You can assign the issue to yourself.

JyotinderSingh replied:
On second thought, maybe it is better to implement LRANGE as a part of this PR itself? It will allow us to verify functionality for this command as well. What do you think?

ParvBudh28 replied:
Sure, I can add implementation for LRANGE as well. My only concern was size of the PR, but that should be okay it seems.

JyotinderSingh commented on integration_tests/commands/async/deque_test.go:
```diff
@@ -442,6 +442,40 @@ func TestLLEN(t *testing.T) {
+		},
+		{
+			name:   "LINSERT after",
+			cmds:   []string{"LINSERT k after v2 e1", "LINSERT k after v1 e2", "LINSERT k after v4 e3"},
```
The test case should also verify the final state of the list by running `LRANGE` at the end.

ParvBudh28 replied:
Ditto as above.

JyotinderSingh commented on internal/eval/deque_test.go:
We should probably include some tests for the newly added methods and structures inside the deque file.

ParvBudh28 replied:
Tests for all public methods are added.

JyotinderSingh commented on internal/eval/deque_test.go:
```diff
@@ -85,6 +87,87 @@ func dequeLPushIntStrMany(howmany int, deq eval.DequeI) {
+	}
+}
+
+func BenchmarkBasicDequeLInsertBefore20(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+		dequeLInsertIntStrMany(20, "before", eval.NewBasicDeque())
+	}
+}
+
+func BenchmarkBasicDequeLInsertBefore200(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+		dequeLInsertIntStrMany(200, "before", eval.NewBasicDeque())
+	}
+}
+
+func BenchmarkBasicDequeLInsertBefore2000(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+		dequeLInsertIntStrMany(2000, "before", eval.NewBasicDeque())
+	}
+}
```
instead of so many cases, just check for a single case with a high number of elements - such as 20k. Same for others.

ParvBudh28 replied:
Done.

JyotinderSingh commented on internal/eval/eval.go:
```diff
@@ -5268,3 +5268,39 @@ func evalJSONSTRAPPEND(args []string, store *dstore.Store) []byte {
+	}
+
+	key := args[0]
+	beforeAfter := strings.ToLower(args[1])
```
There are no checks ensuring that beforeAfter contains a valid value.

JyotinderSingh commented on internal/eval/deque.go:
```diff
@@ -93,8 +95,94 @@ func (q *DequeBasic) LPop() (string, error) {
+		return -1, nil
+	}
+	if beforeAfter != Before && beforeAfter != After {
+		return -1, errors.New("invalid before/after argument")
```
Redis returns the following:
`(error) syntax error`

Please add a unit test for this edge case.

ParvBudh28 replied:
Done.

JyotinderSingh commented on internal/eval/eval_test.go:
```diff
@@ -5733,3 +5734,40 @@ func BenchmarkEvalJSONSTRAPPEND(b *testing.B) {
 	}
 }
+
+func testEvalLINSERT(t *testing.T, store *dstore.Store) {
```
These tests should also validate the final state of the list (wherever applicable).

ParvBudh28 replied:
Done.

JyotinderSingh reviewed (changes requested):
Thanks for this massive effort of implementing LINSERT, @ParvBudh28! I have left some comments.

JyotinderSingh commented on internal/eval/deque.go:
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
+	}
+}
+
+func (q *Deque) insertBeforePivotHelper(pivot, element string, qIterator *DequeIterator, pivotNode *byteListNode) {
```
Add a comment explaining what this method does.

ParvBudh28 replied:
Done.

JyotinderSingh commented on internal/eval/deque.go:
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
+	return elementNode
+}
+
+func (q *Deque) breakPivotNodeAndInsertBefore(qIterator *DequeIterator, pivotNode, elementNode *byteListNode, pivotEntryLen, leftIdx int) *byteListNode {
```
Add a comment explaining what this method does.

ParvBudh28 replied:
Done.

JyotinderSingh commented on internal/eval/deque.go:
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
+	return newNode
+}
+
+func (q *Deque) insertAfterPivotNodeHelper(element string, qIterator *DequeIterator, pivotNode *byteListNode) {
```
Add a comment explaining what this method does.

ParvBudh28 replied:
Done.

JyotinderSingh commented on internal/eval/deque.go:
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
 	return x, nil
 }
 
+func (q *Deque) breakPivotNodeAndInsertAfter(qIterator *DequeIterator, pivotNode *byteListNode, element string) *byteListNode {
```
Add a comment explaining what this method does.

ParvBudh28 replied:
Done.

JyotinderSingh commented on internal/eval/deque.go:
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
+	return newNode
+}
+
+func (q *Deque) updateHeadLInsert(newNode, prevHead *byteListNode, prevLeftIdx int) {
```
Add a comment explaining what this method does.

ParvBudh28 replied:
Done.

JyotinderSingh commented on internal/eval/deque.go:
```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
+	}
+}
+
+func (q *Deque) LInsert(pivot, element, beforeAfter string) (int64, error) {
```
Add a comment explaining what this method does.

ParvBudh28 replied:
Done.

JyotinderSingh commented on internal/eval/eval.go:
Let's add this into store_eval and directly have this supported in the new RESP server (since we are already in process of migrating commands from async to resp).
This will save us a lot of future work.

ParvBudh28 replied:
> Let's add this into store_eval and directly have this supported in the new RESP server (since we are already in process of migrating commands from async to resp). This will save us a lot of future work.

Sure, I can take a look at this tomorrow. Meanwhile it would be great if you could review the additional changes.

ParvBudh28 replied:
This is done.

JyotinderSingh commented on integration_tests/commands/async/deque_test.go:
Once the command is added into store_eval, we should add tests into the resp integration tests directory as well.

JyotinderSingh reviewed (changes requested):
Have added a few more comments based on recent guidelines. Please have a look.

Also, could you please include documentation for these changes as a part of this PR?

JyotinderSingh commented:
@ParvBudh28 Could you please update the PR with the requested changes? The state of the PR is quite close to ideal and it will be good to have it merged before conflicts arise.

ParvBudh28 commented:
> @ParvBudh28 Could you please update the PR with the requested changes? The state of the PR is quite close to ideal and it will be good to have it merged before conflicts arise.

I've made the changes as per the comments. Apologies for the delay. Was caught up during the week.

JyotinderSingh reviewed (changes requested):
The changes look good, based on the new PR guidelines please also update the documentation for this command under the docs directory.

ParvBudh28 commented:
> The changes look good, based on the new PR guidelines please also update the documentation for this command under the docs directory.
> 
> Please also fix the merge conflicts and the linter errors.

Done. I've fixed the linter issues and added the documentation as well.

apoorvyadav1111 commented:
Hi @ParvBudh28, Please find the checklist for new commands/ migration PRs. We are moving away from async server to resp and new commands are added to store_eval.go instead of eval.go.


- [x] *moved* code from eval.go to store_eval.go
- [x] *updated* cmd meta in eval, worker & server
- [x] *moved* tests from async to resp
- [x] *added/updated* tests in http and websocket
- [x] *added/updated* unit tests in eval folder
- [x] *audited* documentation and updated as per new standards
- [x] *no changes* in legacy async code / eval.go file
- [x] *rebased* with master
- [x] *verified* the code in local

Please checkout the PR #1117 on migration/adding new commands and related artifacts.

Thanks

ParvBudh28 commented:
> Hi @ParvBudh28, Please find the checklist for new commands/ migration PRs. We are moving away from async server to resp and new commands are added to store_eval.go instead of eval.go.
> 
> * [ ]  _moved_ code from eval.go to store_eval.go
> * [ ]  _updated_ cmd meta in eval, worker & server
> * [ ]  _moved_ tests from async to resp
> * [ ]  _added/updated_ tests in http and websocket
> * [ ]  _added/updated_ unit tests in eval folder
> * [x]  _audited_ documentation and updated as per new standards
> * [ ]  _no changes_ in legacy async code / eval.go file
> * [ ]  _rebased_ with master
> * [ ]  _verified_ the code in local
> 
> Please checkout the PR #1117 on migration/adding new commands and related artifacts.
> 
> Thanks

Hi @apoorvyadav1111, I've moved the code away from eval.go to store_eval.go and from async to resp. Please check and let me know if there are any other action items here.

apoorvyadav1111 commented on internal/eval/commands.go:
```diff
@@ -1206,6 +1206,45 @@ var (
+			-1 when the pivot wasn't found.
+		`,
+		Eval:     evalLINSERT,
+		Arity:    4,
```
as per redis command info it is 5, I guess they count the command name as well. 
```suggestion
		Arity:    5,
```

ParvBudh28 replied:
Done.

apoorvyadav1111 commented on internal/eval/commands.go:
```diff
@@ -1206,6 +1206,45 @@ var (
+			Array reply: a list of elements in the specified range, or an empty array if the key doesn't exist.
+		`,
+		Eval:     evalLRANGE,
+		Arity:    3,
```
as per redis command info it is 4, I guess they count the command name as well. 
```suggestion
		Arity:    4,
```

ParvBudh28 replied:
Done.

apoorvyadav1111 commented on internal/eval/commands.go:
```diff
@@ -1206,6 +1206,45 @@ var (
+			0 when the key doesn't exist.
+			-1 when the pivot wasn't found.
+		`,
+		Eval:     evalLINSERT,
```
```suggestion
		NewEval:     evalLINSERT,
		IsMigrated: true,
```

apoorvyadav1111 commented on internal/eval/commands.go:
```diff
@@ -1206,6 +1206,45 @@ var (
+		Returns:
+			Array reply: a list of elements in the specified range, or an empty array if the key doesn't exist.
+		`,
+		Eval:     evalLRANGE,
```
```suggestion
		IsMigrated: true,
		NewEval:     evalLRANGE,
```

ParvBudh28 replied:
Done.

apoorvyadav1111 commented on internal/eval/eval_test.go:
We need to migrate to using EvalResonse here instead of []byte

ParvBudh28 replied:
Done.

apoorvyadav1111 commented on internal/eval/store_eval.go:
Please refer to get, set command in store_eval on return type of new/ migrated commands

ParvBudh28 replied:
Done.

apoorvyadav1111 reviewed (changes requested):
Hi @ParvBudh28, I have left some comments which are primarily related to our new architecture. Please update cmd_meta.go files, any documentation in docs folder (refer: docs/command_docs_template.md), new tests in HTTP and websocket folder to close this PR. 

It can be a lot of changes, however, we can always assist you. Please reach out to us on discord if you have any questions.

Apoorv

JyotinderSingh commented:
@ParvBudh28 The PR is almost ready to be merged, could you please address the last few comments here?

ParvBudh28 commented on integration_tests/commands/resp/deque_test.go:
```diff
@@ -0,0 +1,93 @@
+	"net"
+	"testing"
+
+	"gotest.tools/v3/assert"
```
Done.

ParvBudh28 commented:
> @ParvBudh28 The PR is almost ready to be merged, could you please address the last few comments here?

Hey Apologies for the delay, I was traveling last week. I have addressed the comments and updated the PR.

JyotinderSingh reviewed (approved):
Thanks for addressing all the comments @ParvBudh28! The changes look great. Approved.

ParvBudh28 commented:
> Thanks for addressing all the comments @ParvBudh28! The changes look great. Approved.

 @apoorvyadav1111 Is your approval also required to close this PR ?

apoorvyadav1111 commented:
Hi @ParvBudh28, Please rebase your branch. There are some conflicts which are blocking this merge.

Thanks,
Apoorv

ParvBudh28 commented:
> Hi @ParvBudh28, Please rebase your branch. There are some conflicts which are blocking this merge.
> 
> Thanks, Apoorv

@apoorvyadav1111 Done.

JyotinderSingh commented:
Thanks for the massive effort on this PR @ParvBudh28!
